src/windows/Button.tsx

```diff
@@ -0,0 +1,211 @@
+    }
+
+    private _onKeyDown = (e: React.SyntheticEvent<any>): void => {
+
+        if (!this.props.disabled) {
+            let keyEvent = EventHelpers.toKeyboardEvent(e);
+            if (this.props.onKeyPress) {
+                this.props.onKeyPress(keyEvent);
+            }
+
+            if (this.props.onPress) {
+                let key = keyEvent.keyCode;
+                // ENTER triggers press on key down
+                if (key === KEY_CODE_ENTER) {
+                    this.props.onPress(keyEvent);
+                    return;
+                }
+            }
+        }
+    }
+
+    private _onKeyUp = (e: React.SyntheticEvent<any>): void => {
+
+        let keyEvent = EventHelpers.toKeyboardEvent(e);
+        if (keyEvent.keyCode === KEY_CODE_SPACE) {
+            if (!this.props.disabled && this.props.onPress) {
+                this.props.onPress(keyEvent);
+            }
+        }
+    }
+
+    private _onMouseEnter = (e: React.SyntheticEvent<any>) => {
+        this._isMouseOver = true;
+        this._onHoverStart(e);
+    }
+
+    private _onMouseLeave = (e: React.SyntheticEvent<any>) => {
+        this._isMouseOver = false;
+        this._onHoverEnd(e);
+    }
+
+    // When we get focus on an element, show the hover effect on the element.
+    // This ensures that users using keyboard also get the similar experience as mouse users for accessibility.
+    private _onFocus = (e: React.SyntheticEvent<any>): void => {
+        this.onFocus();
+
+        this._isFocusedWithKeyboard = _isNavigatingWithKeyboard;
+        this._onHoverStart(e);
+
+        if (this.props.onFocus) {
+            this.props.onFocus(EventHelpers.toFocusEvent(e));
+        }
+    }
+
+    private _onBlur = (e: React.SyntheticEvent<any>): void => {
+        this._isFocusedWithKeyboard = false;
+        this._onHoverEnd(e);
+
+        if (this.props.onBlur) {
+            this.props.onBlur(EventHelpers.toFocusEvent(e));
+        }
+    }
+
+    private _onHoverStart = (e: React.SyntheticEvent<any>) => {
+        if (!this._isHoverStarted && (this._isMouseOver || this._isFocusedWithKeyboard)) {
+            this._isHoverStarted = true;
+
+            if (this.props.onHoverStart) {
+                this.props.onHoverStart(e);
+            }
+        }
+    }
+
+    private _onHoverEnd = (e: React.SyntheticEvent<any>) => {
+        if (this._isHoverStarted && !this._isMouseOver && !this._isFocusedWithKeyboard) {
+            this._isHoverStarted = false;
+
+            if (this.props.onHoverEnd) {
+                this.props.onHoverEnd(e);
+            }
+        }
+    }
+
+    onFocus() {
+        // Focus Manager hook
+    }
+
+    getTabIndex(): number | undefined {
+        // Button defaults to a tabIndex of 0
+        // Focus Manager may override this
+        return this.props.tabIndex || 0;
+    }
+
+    updateNativeTabIndex(): void {
+        if (this._focusableElement) {
+            let tabIndex: number | undefined = this.getTabIndex();
```

Comment:
>: number | undefined [](start = 24, length = 20)

Is it preferred to allow TypeScript to infer the type in cases like this?

Reply:
I don't know, I've seen either way. Can be argued that it's clearer (to the viewer) this way.

---
In reply to: [159337210](https://github.com/Microsoft/reactxp/pull/404#discussion_r159337210) [](ancestors = 159337210)